### PR TITLE
Small Operator Nodes

### PR DESCRIPTION
**Pull request by mddvisu**
Added "Small Operator" nodes to script canvas which are meant to perform small operations on data without taking up much space in the graph. As of now these nodes do not parse, so they are hidden by default. To show these nodes in the palette, flip the ed_showSmallOperatorNodes cvar to true, and they will appear in the node palette under Math/Small Operators.

These nodes are created in a more data-driven way, keeping the future direction of ScriptCanvas in mind. Instead of the information for these nodes being stored in an inherited class of Node, these nodes are created through user data, allowing for many different small operator nodes to be created in very little time.

Small Operator Nodes

### CODE REVIEW

File: Gems/GraphCanvas/Code/Source/Components/Nodes/General/GeneralNodeLayoutComponent.cpp [hunk 1]
```diff
@@ -97,6 +100,18 @@ namespace GraphCanvas
     
     void GeneralNodeLayoutComponent::OnStyleChanged()
     {
+        if (!m_hasSwitchedToHorizontalLayout)
+        {
+            Styling::StyleHelper style(GetEntityId());
+            Qt::Orientation layoutOrientation = style.GetAttribute(Styling::Attribute::LayoutOrientation, Qt::Vertical);
```
Comment (carlitosan) [inline Gems/GraphCanvas/Code/Source/Components/Nodes/General/GeneralNodeLayoutComponent.cpp:106]: Is this new member bool necessary? Can you query the current LayoutOrientation on demand?

File: Gems/ScriptCanvas/Code/Editor/Nodes/NodeCreateUtils.h [hunk 2]
```diff
@@ -26,6 +28,22 @@ namespace ScriptEvents
...
+
+    // User defined data for small operator nodes
+    struct SmallOperatorCreationData
+    {
+        AZ_RTTI(SmallOperatorCreationData, "{A7DE9ECF-81F9-4A03-B3E4-3510591A50BB}");
+        AZ_CLASS_ALLOCATOR(SmallOperatorCreationData, AZ::SystemAllocator, 0);
+
+        static void Reflect(AZ::ReflectContext* reflectContext);
+
+        virtual ~SmallOperatorCreationData() = default;
```
Comment (carlitosan) [inline Gems/ScriptCanvas/Code/Editor/Nodes/NodeCreateUtils.h:40]: You don't need this destructor definition, or for it to be virtual.

File: Gems/ScriptCanvas/Code/Editor/Nodes/NodeCreateUtils.h [hunk 3]
```diff
@@ -26,6 +28,22 @@ namespace ScriptEvents
 
 namespace ScriptCanvasEditor::Nodes
 {
+
+    // User defined data for small operator nodes
+    struct SmallOperatorCreationData
+    {
+        AZ_RTTI(SmallOperatorCreationData, "{A7DE9ECF-81F9-4A03-B3E4-3510591A50BB}");
```
Comment (carlitosan) [inline Gems/ScriptCanvas/Code/Editor/Nodes/NodeCreateUtils.h:35]: Use AZ_TYPE_INFO

File: Gems/ScriptCanvas/Code/Editor/Nodes/NodeCreateUtils.cpp [hunk 4]
```diff
@@ -237,6 +253,52 @@ namespace ScriptCanvasEditor::Nodes
...
+        AZ::Entity* nodeEntity{ aznew AZ::Entity };
+        nodeEntity->Init();
+        nodeEntity->SetName(node->GetNodeName());
+        nodeEntity->AddComponent(node);
+        ScriptCanvas::GraphRequestBus::Event(scriptCanvasId, &ScriptCanvas::GraphRequests::AddNode, nodeEntity->GetId());
+
+        node->AddSlot(inputPin, true);
+        node->AddSlot(outputPin, true);
+
+        node->SetNodeStyle(GraphCanvas::Styling::Elements::Small);
```
Comment (carlitosan) [inline Gems/ScriptCanvas/Code/Editor/Nodes/NodeCreateUtils.cpp:290]: Pipe this request through the DataDrivenNode pipeline, and then remove references to any SmallOperatorNode classes, names. Just use the data-driven pipeline that you have provided.

File: Gems/ScriptCanvas/Code/Include/ScriptCanvas/Core/Node.cpp [hunk 5]
```diff
@@ -2963,27 +2967,60 @@ namespace ScriptCanvas
 
     AZStd::string Node::GetNodeName() const
     {
-        AZ::SerializeContext* serializeContext = nullptr;
-        AZ::ComponentApplicationBus::BroadcastResult(serializeContext, &AZ::ComponentApplicationRequests::GetSerializeContext);
-
-        if (serializeContext)
+        if (m_name == "")
```
Comment (carlitosan) [inline Gems/ScriptCanvas/Code/Include/ScriptCanvas/Core/Node.cpp:2970]: if (m_name.empty())

File: Gems/ScriptCanvas/Code/Include/ScriptCanvas/Core/Node.cpp [hunk 6]
```diff
@@ -2963,27 +2967,60 @@ namespace ScriptCanvas
...
+    {
+        m_toolTip = toolTip;
+    }
+
+    void Node::SetNodeStyle(const AZStd::string& nodeStyle)
+    {
+        m_nodeStyle = nodeStyle;
+    }
+
+    void Node::SetNodenodeLexicalId(const AZ::Crc32& nodeLexicalId)
```
Comment (carlitosan) [inline Gems/ScriptCanvas/Code/Include/ScriptCanvas/Core/Node.cpp:3021]: SetNodeLexicalId

File: Gems/ScriptCanvas/Code/Editor/View/Widgets/NodePalette/NodePaletteModel.h [hunk 7]
```diff
@@ -59,6 +61,17 @@ namespace ScriptCanvasEditor
         AZStd::string m_tooltip;
     };
 
+    struct RegisterSmallOperatorNodeInformation
```
Comment (carlitosan) [inline Gems/ScriptCanvas/Code/Editor/View/Widgets/NodePalette/NodePaletteModel.h:64]: This all can just be the DataDrivenTypeInfo, and doesn't need to refer to small operators

**Review by carlitosan — CHANGES_REQUESTED** [review]

**Review by lsemp3d — APPROVED** [review]

**Review by carlitosan — APPROVED** [review]

**Pull request merged**